src/app/ReadClient.cpp

```diff
@@ -0,0 +1,326 @@
+                    eventPathBuilder.EventId(eventPath.mEventId);
+
+                eventPathBuilder.EndpointId(eventPath.mEndpointId).ClusterId(eventPath.mClusterId).EndOfEventPath();
+                SuccessOrExit(eventPathBuilder.GetError());
+            }
+
+            eventPathListBuilder.EndOfEventPathList();
+            SuccessOrExit(eventPathListBuilder.GetError());
+
+            if (outReadParam.mReadRequestPrepareNeeded.mEventNumber != 0)
+            {
+                request.EventNumber(outReadParam.mReadRequestPrepareNeeded.mEventNumber);
+            }
+        }
+
+        request.EndOfReadRequest();
+        SuccessOrExit(request.GetError());
+
+        err = writer.Finalize(&msgBuf);
+        SuccessOrExit(err);
+
+        VerifyOrExit(msgBuf->EnsureReservedSize(System::PacketBuffer::kDefaultHeaderReserve), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    }
+
+    ClearExistingExchangeContext();
+
+    // Create a new exchange context.
+    // TODO: temporary create a SecureSessionHandle from node id, will be fix in PR 3602
+    mpExchangeCtx = mpExchangeMgr->NewContext({ aNodeId, 0, aAdminId }, this);
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
+    mpExchangeCtx->SetResponseTimeout(kIM_MSG_TIMEOUT);
+
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReadRequest, std::move(msgBuf),
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    SuccessOrExit(err);
+    MoveToState(kState_ReadRequestSending);
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadClient::OnMessageReceived(Messaging::ExchangeContext * apEc, const PacketHeader & aPacketHeader,
+                                   const PayloadHeader & aPayloadHeader, System::PacketBufferHandle aPayload)
+{
+    // Assert that the exchange context matches the client's current context.
+    // This should never fail because even if SendCommandRequest is called
+    // back-to-back
+
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrDie(apEc == mpExchangeCtx);
+
+    ClearExistingExchangeContext();
+
+    err = ProcessReportData(std::move(aPayload));
+    SuccessOrExit(err);
+
+exit:
+    Reset();
+    return;
+}
+
+CHIP_ERROR ReadClient::ClearExistingExchangeContext()
+{
+    // Discard any existing exchange context. Effectively we can only have one Echo exchange with
+    // a single node at any one time.
+    if (mpExchangeCtx != nullptr)
+    {
+        mpExchangeCtx->Abort();
+        mpExchangeCtx = nullptr;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReadClient::ProcessReportData(System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    InEventParam inParam;
+    OutEventParam outParam;
+    ReportData::Parser report;
+
+    bool isEventListPresent = false;
+    bool suppressResponse   = false;
+
+    chip::System::PacketBufferTLVReader reader;
+
+    {
+        inParam.Clear();
```

Comment:
Constructors should build a cleared out object from the start.

Reply:
Done, thanks